subsys/net/lib/lwm2m/lwm2m_obj_access_control.c

```diff
@@ -0,0 +1,433 @@
+	if (idx < 0) {
+		LOG_DBG("Cannot remove access control for /%d/%d - not found", obj_id, obj_inst_id);
+		return;
+	}
+
+	m_ac_data[idx].obj_id = 0;
+	m_ac_data[idx].obj_inst_id = 0;
+	m_ac_data[idx].ac_owner = 0;
+	for (int i = 0; i < MAX_SERVER_COUNT + 1; i++) {
+		m_ac_data[idx].acl[i] = 0;
+	}
+	lwm2m_delete_obj_inst(AC_OBJ_ID, idx);
+}
+
+void access_control_remove_obj(uint16_t obj_id) { access_control_remove(obj_id, OBJ_LVL_MAX_ID); }
+
+static bool check_acl_table(uint16_t obj_id, uint16_t obj_inst_id, uint16_t short_server_id,
+			    uint16_t access)
+{
+	/* Get the index of the ac instance regarding obj_id and obj_inst_id */
+	int idx = obj_inst_to_index(obj_id, obj_inst_id);
+
+	if (idx < 0) {
+		LOG_DBG("Access control for obj_inst /%d/%d not found", obj_id, obj_inst_id);
+		return false;
+	}
+
+	uint16_t access_rights = 0;
+	uint16_t default_rights = 0;
+	bool server_has_acl = false;
+
+	for (int i = 0; i < MAX_SERVER_COUNT + 1; i++) {
+		int res_inst_id = res_inst[idx][ACCESS_CONTROL_ACL_ID + i].res_inst_id;
+		/* If server has access or if default exist */
+		if (res_inst_id == short_server_id) {
+			access_rights |= m_ac_data[idx].acl[i];
+			server_has_acl = true;
+		} else if (res_inst_id == 0) {
+			default_rights |= m_ac_data[idx].acl[i];
+		}
+	}
+
+	if (server_has_acl) {
+		return (access_rights & access) == access;
+	}
+	/* Full access if server is the ac_owner and no acl is specified for that server */
+	if (m_ac_data[idx].ac_owner == short_server_id) {
+		return true;
+	}
+
+	/* Return default rights */
+	return (default_rights & access) == access;
+}
+
+int access_control_check_access(uint16_t obj_id, uint16_t obj_inst_id, uint16_t server_obj_inst,
+				uint16_t lwm2maccess, bool bootstrap_mode)
+{
+#if defined(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP)
+	if (bootstrap_mode) {
+		return 0; /* Full access for bootstrap servers */
+	}
+#else
+	ARG_UNUSED(bootstrap_mode);
+#endif
+	/* If ac_obj not created */
+	if (!ac_obj.fields) {
+		return 0;
+	}
+	uint16_t access = lwm2mperm_to_acperm(lwm2maccess);
+	int short_server_id = lwm2m_server_get_ssid(server_obj_inst);
+
+	if (short_server_id < 0) {
+		LOG_ERR("No server obj %u exist", server_obj_inst);
+		return -EACCES;
+	}
+
+	if (obj_id == AC_OBJ_ID) {
+		switch (access) {
+		case READ:
+			return 0;
+		case ACEXEC:
+		case DELETE:
+		case CREATE:
+			return -EPERM; /* Method not allowed */
+		case WRITE:	       /* Only ac_owner can write to ac_obj */
+			for (int index = 0; index < ARRAY_SIZE(inst); index++) {
+				if (inst[index].obj && inst[index].obj_inst_id == obj_inst_id) {
+					if (m_ac_data[index].ac_owner == short_server_id) {
+						return 0;
+					}
+				}
+			}
+			return -EACCES;
+
+		default:
+			return -EACCES;
+		}
+	}
+
+	/* only DICOVER, WRITE_ATTR and CREATE allowed on object*/
+	if (obj_inst_id == OBJ_LVL_MAX_ID) {
+		if (access == DISCOVER || access == WRITE_ATTR) {
+			return 0;
+		}
+
+		if (access != CREATE) {
+			return -EACCES;
+		}
+		return 0;
```

Comment:
Couldn't this be covered by a single `if`?
```suggestion
		if (access == DISCOVER || access == WRITE_ATTR || access == CREATE) {
			return 0;
		}

		return -EACCES;
```

Reply:
Yes, but I made an error. The return on line 301 should not have been there.